### PR TITLE
django-field-audit: initial implementation

### PR DESCRIPTION
**Pull request by joxl**
This is the initial implementation of this library.

See the [design document](https://docs.google.com/document/d/1-9CGoeJD0iPozqO7Wp1KwqerxhPq6HCjBZt2kkEQpkE/edit#) for high level specifications that drove the design of this plugin.

There are a few key features that are not yet completed which will be worked on during code review, including:

- "Backfill" migration utilities for writing initial audit records for models that existed prior to this app being enabled.
- Add support for `QuerySet` write operations (`update()`, etc).
- High-level library documentation (see below).

This library does not yet have high-level usage docs (this is near the top of the list of next steps). I decided to push the library without them for now. The HQ dimagi/commcare-hq#31732 PR demonstrates usage until the full docs are written.

This Django app was inspired by the [field_history](https://pypi.org/project/django-field-history/) app, and is designed to do a very similar task, with some key differences. Some implementation details of the `field_history` app make it incompatible with HQ auditing needs:
- implementation was designed with foreign keys to the Django `AUTH_USER_MODEL`, which prohibit auditing "changed by" information besides the built-in Django authentication models (HQ needs to track more complex user models).
- auth model foreign key uses `ON DELETE CASCADE` references, meaning that deleting a user will also delete the entire change history created by that user.
- only overrides the `<model_instance>.save()` method on tracked models, and doesn't record delete events (HQ needs to track `delete()`s as well).
- every "field change" record is a completely unique object with no reference to the "event" that caused it, making it difficult to tell which field changes are associated with which "change actions" (e.g. `instance.save()`) on a model.
- does not save the _previous_ value for a field change event, which makes using the library for field audit purposes more challenging (requires extra queries to determine what the value was before it changed).
- saves transient tracking information on the model instance itself via `setattr(instance, f"_{field_name}", value)`, which is fragile and will silently overwrite existing private field attributes.
- field values are JSON-serialized and stored in a `CharField` rather than a `JSONField`.

I did attempt to fork the `field_history` project and update/modernize it in order to conform with HQ auditing constraints. The fork attempt quickly turned out to be far more overhead than justifiable for the sake of maintaining the "shell" of an existing project. The `field_history` app is ~350 lines of python code (~900 with tests). Given the magnitude of changes needed for HQ, I decided to roll a new app using the former app as inspiration.

### CODE REVIEW

File: README.md
Comment (snopoke) [inline README.md:100]: Curious why we are using nose which is no longer maintained. Copying this note from the docs:

> Nose has been in maintenance mode for the past several years and will likely cease without a new person/team to take over maintainership. New projects should consider using [Nose2](https://github.com/nose-devs/nose2), [py.test](http://pytest.org/), or just plain unittest/unittest2.

I would recommend pytest over nose2 (I believe nose2 already recommends using pytest)

Comment (snopoke) [reply README.md:100]: I see that's already a todo

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:45]: stray comment

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:229]: statement has no effect

**Review by snopoke — COMMENTED** [review]
I'll set aside some time to review more closely this week.

File: field_audit/models.py
Comment (gherceg) [inline field_audit/models.py:124]: Is this intentional?

File: field_audit/models.py
Comment (gherceg) [inline field_audit/models.py:171]: nit
```suggestion
            to attach for later auditing
```

File: field_audit/models.py
Comment (gherceg) [inline field_audit/models.py:151]:
```suggestion
        :param field_name: an attribute name or Python dot-path of attribute
```

**Review by gherceg — COMMENTED** [review]
Just minor comments on first read

File: README.md
Comment (joxl) [reply README.md:49]: Yeah, there is room for improvement here. Omitting the middleware does _not_ break the app (this is an intentional design feature).  The key is that the middleware is required if you want `changed_by` auditing information to come from _requests_, definitely some better wording is needed. I'll update this.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:124]: Yes, this is an artifact of me not making up my mind on the whole `AutoField` thing. Would be an interesting discussion to have, I think.

File: tests/test_models.py
Comment (millerdev) [inline tests/test_models.py:484]: same as above

File: field_audit/field_audit.py
Comment (dannyroberts) [inline field_audit/field_audit.py:83]: You probably know this and maybe got the pattern from there, but commcare-hq has a very similar utility: https://github.com/dimagi/commcare-hq/blob/579b52f1037d30c8e9184214f7278bb1d6a8b9d2/corehq/util/global_request/api.py#L6. It feels a little funny to have to do this twice on every request and add another layer of middleware on every request, but I don't think practically it would cause any problems. I guess the alternative would be to make the calling project pass in a `get_request` method in the settings or something, which feels even odder. So no actual suggestions here, but I wanted to point that out in case it made you think of something.

**Review by calellowitz — COMMENTED** [review]
Left a few questions, but haven't had a chance yet to look deeply at the actual implementation. A few bigger picture questions.

Is there a high level explanation of what this does? That would make reviewing the code a lot easier, and especially determining whether a given behavior is intentional.

Is there a plan to do performance testing on this, to see how it affects request timings on pages that touch these models?

File: field_audit/models.py
Comment (calellowitz) [reply field_audit/models.py:124]: > Would be an interesting discussion to have, I think
Curious what this discussion is?

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:23]: My thought process was "I don't know what _every_ db engine string looks like". I could go ether way on this. As-is it's not particularly great. I figured if this became a problem in the future then a bug report would likely clarify a wider scenario that needs to be handled.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:124]: Prior to Django 3.2, if you didn't explicitly configure a `primary_key` field, Django would implicitly add one for you using `AutoField`. Since 3.2, if you don't specify a primary key column (i.e. you use the "implicit behavior" _and_ you don't specify either `YourAppConfig.default_auto_field` or `settings.DEFAULT_AUTO_FIELD`), Django will emit warnings when you call `makemigrations`.  I didn't want this app to **hard define** which AutoField a downstream project must use (`AutoField` vs. `BigAutoField`), so I added a (still undocumented) app setting `FIELD_AUDIT_AUTO_FIELD` to provide downstream projects the ability to choose. I'm still undecided on the merits of this. On one hand it might be overkill (just hard-require `BigAutoField`), on the other hand since Django doesn't force this decision, it seems reasonable to leave it up to downstream "consumers".

File: field_audit/auditors.py
Comment (joxl) [reply field_audit/auditors.py:106]: The problem is that the current implementation of the auditor chain "behavior" doesn't provide a mechanism for auditor objects _in_ the chain to be aware of each other. As such, if a "request auditor" (one that knows how to audit an HTTP request) isn't able to extract audit information about a specific request, it just returns `None`: ignorant to whether it's the _last_ "request auditor" or if there's another (more specialized one) that will know what to do with the request.

With that context, in the current implementation using the default audit chain (e.g. `[RequestAuditor(), SystemUserAuditor()]`), if an unauthenticated request is run down the chain and the `SystemUserAuditor` does _not_ return `None`, the resulting `changed_by` value will report the Django WSGI server's OS user, which is probably not an an "expected value" in that scenario.

This does feel a bit like I'm "guessing in the face of ambiguity".  Perhaps it would be better to implement another auditor in the chain (something like `UnauthenticatedRequestAuditor`) that returns a static value (like `{"user_type": None}`) and document why you should consider using it you define your own auditor chain and wish to short-circuit on unauthenticated web requests? In this case it might be better to also implement a constant "I don't know" value (e.g. `BaseAuditor.NOT_FOUND`) so an auditor in the chain can return `None` meaning "I know the `changed_by` value is actually `None`.

Comment (joxl) [reply field_audit/auditors.py:106]: Offlined with @millerdev about this. I will make an update with a cleaner solution.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:323]: In this scenario the behavior varies (depending on other args), which is why I've explicitly prohibited it. If you pass `FieldChange.MISSING` when either `(is_create or is_delete) == True`, then the eventual "change check" [here](https://github.com/dimagi/django-field-audit/blob/4808788a04467879e8da546895c8dbcc3b732d3c/field_audit/models.py#L331) (`if old_value == new_value:`) will be `False`, resulting in a "calculated no change" scenario which will not create a new `AuditEvent` record. This would be an error condition because `is_create` _or_ `is_delete` _always_ implies there has been a change. Perhaps this function could be refactored to handled this in a more elegant way?

This could be made more efficient (which would result in the JSON serializer to raising a `ValueError` later on if you pass `value=FieldChange.MISSING`):

```diff
-        if value is missing:
-            raise ValueError("'value' cannot be MISSING")
-        if (is_create or is_delete) and init_value is not missing:
-            raise ValueError(
-                "'init_value' is mutually exclusive with 'is_create' and "
-                "'is_delete'"
-            )
+        if is_create or is_delete:
+            if value is missing:
+                raise ValueError("'value' cannot be MISSING")
+            if init_value is not missing:
+                raise ValueError(
+                    "'init_value' is mutually exclusive with 'is_create' and "
+                    "'is_delete'"
+                )
```

File: CHANGELOG.md
Comment (joxl) [reply CHANGELOG.md:6]: Because I burned version `0.1` by accidentally uploading it to my personal PyPi account (and you can't push the same version of the same package more than once, even if the original package has been deleted. :facepalm: In order to upload the package to the Dimagi account, I had to bump the version (in hindsight I assume `0.1.1` would have been sufficient, but I happened to try `0.2` because I didn't actually expect it to work).

File: field_audit/field_audit.py
Comment (joxl) [reply field_audit/field_audit.py:65]: Yes, in fact I'm considering adding this even before _this_ version goes live in HQ. I know that `update` doesn't call `save`, and I assume most (if not all) of the `QuerySet` methods don't call any methods on the model itself.

File: README.md
Comment (joxl) [reply README.md:147]: No, it means if you try to do something clever by decorating the same model twice in the same runtime, it will fail the second time. Like this:

```python
@audit_fields("tail_number")
class Aircraft(models.Model):
    tail_number = models.CharField(max_length=32, unique=True)
    operated_by = models.CharField(max_length=64)

# ^ nothing wrong here

# ...

Aircraft = audit_fields("operated_by")(Aircraft)  # <-- currently, this will raise
```
This is not a priority, and may never get changed.

File: field_audit/field_audit.py
Comment (joxl) [reply field_audit/field_audit.py:83]: Oh, I didn't know that HQ does this. This is the same pattern that the `field_history` app uses (albeit, that app does this in a more subtle way). I would actually like to get rid of this completely, but it's not immediately clear how to do it any "better". The `django.core.signals.request_started` would suffice for this app (to avoid middleware). I don't know if there are any performance benefits to alternative, my naive assumption is "probably not". Even then, I would still need to use a global variable to store the request (for auditors to reference), unless I want to attach the audit state _to_ the request itself (disclosure: I didn't investigate this solution at all). If you have more thoughts on any of my rambling here, I'm open to further conversation.

File: README.md
Comment (joxl) [reply README.md]: Changed in 8fc7d27f.

File: field_audit/migrations/0002_alter_auditevent_changed_by.py
Comment (AmitPhulera) [inline field_audit/migrations/0002_alter_auditevent_changed_by.py:1]: Great progress here. 

I really enjoyed going through your implementation. 😄 

A small suggestion, it would be nice if we can have this migration in the initial migration file too.

File: field_audit/models.py
Comment (snopoke) [reply field_audit/models.py:124]: For a library like this I think it makes sense to use `BigAutoField` rather than leaving it to the consumer. It makes sense for Django to leave it for the consumer since Django doesn't know what the model will do or how many records to expect etc. Since we DO know that for this model there is less need to leave the decision open.

**Comment by millerdev** [comment]
Would make sense to eliminate the `FieldChange` model and instead store the data in a `jsonb` field of the `AuditEvent` table? The structure of `AuditEvent.changes` would be `{<field_name>: {"old": ..., "new": ...}}`. This would have the benefit of making inserts (the most common operation) very simple and fast (one row per change), and the `changes` field can be [indexed for efficient querying](https://www.postgresql.org/docs/current/datatype-json.html#JSON-INDEXING). Concerns about update performance of large JSON fields can be ignored since the data never changes after it is created.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:124]: Thanks for that feedback @snopoke, I'm going to rip this part out and just explicitly use `BigAutoField`.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py]: Changed in 7898b0a.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py]: Changed in f122828.

File: field_audit/migrations/0002_alter_auditevent_changed_by.py
Comment (joxl) [reply field_audit/migrations/0002_alter_auditevent_changed_by.py]: Migrations are now squashed.

File: field_audit/migrations/0001_initial.py
Comment (snopoke) [inline field_audit/migrations/0001_initial.py:29]: What kind of queries is this index intended to support? Wondering if it needs to be a full GIN index or if it could using `json_path_ops` or even be a B-Tree index if we only need to know if a specific key is present in the data.

File: CHANGELOG.md
Comment (snopoke) [inline CHANGELOG.md:3]: Is this level of detail necessary considering this code is still in PR and unreleased? Just wondering.

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:14]: where is `sqlite` used? It seems like tests run on PostgreSQL. If this is just for general compatibility it seems like a fair amount of added complexity for an unneeded use case (so far).

Also if we keep sqlite you'll want to add a note about JSON support: https://code.djangoproject.com/wiki/JSON1Extension

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:52]: this query makes me think that the `changed_by` field expects a certain structure. The docstring for the auditor just says `changed_by() method that returns a JSON-serializable object`. Perhaps that should be constrained to a dataclass if the code is going to rely on specific field in that field.

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:120]: Is this for mocking in tests? Docstring could be useful

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:126]: if we are storing a class path, what happens if a class get's moved or renamed? Perhaps the decorator could also provide a 'name' for the model that will remain consistent?

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:200]: Should `request` be optional? Seems like you'd want to be able to audit changes that happen outside of a request e.g. shell / management command. At least the docstring should note that it can be `None`

File: field_audit/migrations/0001_initial.py
Comment (joxl) [reply field_audit/migrations/0001_initial.py:29]: Ah, I added this and completely forgot about it (I meant to remove it). I did some testing with indexes into `jsonb` columns. We don't want an index on the `delta` column. There is no known audit query "practice" that requires it: that is, the "expected" queries on this table will be something like:

```python
AuditEvent.objects.filter(object_class_path='app.models.Thing', changed_by__username='someone')
```

When considering HQ specifically (which adds an `HQAuditor` for capturing `changed_by["domain"]`, the "happy path" queries are expected to look something like:

```python
AuditEvent.objects.filter(object_class_path='app.models.Thing', changed_by__domain='the-things')
```

This means that HQ will want to add its own index (perhaps a `GIN`?). For example:

```sql
CREATE INDEX field_audit_auditevent_changed_by_domain
  ON field_audit_auditevent
  USING GIN (changed_by->'domain' jsonb_path_ops);
```
Still thinking on this, a `HASH` index might be better because `changed_by->'username'` (and other `changed_by` keys like `domain`) are not nested objects (the values are just strings), so a `HASH` index might be better suited here. I'll be testing this before I add anything. Feedback is very welcome.

Comment (joxl) [reply field_audit/migrations/0001_initial.py]: Index removed in c5d1300.

File: field_audit/migrations/0001_initial.py
Comment (joxl) [reply field_audit/migrations/0001_initial.py:20]: > Update: Looking at the manager methods below I'm almost certain this should be indexed and that the data needs to be structured in a specific way.

Correct, adding indexes on this column would optimize for fast "default manager" filters (the ones defined below). That being said, indexes come with an insert performance penalty, so I've left them out for now. I'm open to more conversation on this, but my current position is that the value of the indexes depends heavily on how you want to query this table. As such, I've left them out until we better define exactly which queries we need performance on.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:13]: Developing this library doesn't require PostgreSQL, and running tests locally uses SQLite. The GitHub Workflows tests use PostreSQL (because support for that is very important), but local tests don't need it.

File: tests/django_setup.py
Comment (joxl) [reply tests/django_setup.py:14]: Django's `django.db.connection.creation.create_test_db(...)` function can prompt for user input, which is problematic if it happens when invoked by the `nose` test runner because by default, `nose` captures STDOUT. This ensures that if Django prompts, the message cannot be obscured by nose's STDOUT capture.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:13]: > Also if we keep sqlite you'll want to add a note about JSON support: https://code.djangoproject.com/wiki/JSON1Extension

Hmm, I haven't had any trouble with SQLite (running tests). Perhaps that extension is enabled on my system and I don't know it? I'll look further into this and add a note if it turns out SQLite isn't supported as-is.

File: field_audit/models.py
Comment (snopoke) [reply field_audit/models.py:51]: That may be true but then you don't get some of the 'out of box' functionality which seems weird. It should at least be very clearly documented what structure is required in order to get the default queries to work.

File: field_audit/migrations/0001_initial.py
Comment (snopoke) [reply field_audit/migrations/0001_initial.py:20]: Surely we have a pretty clear idea of how we are going to query this table already? I certainly expect that not being able to filter efficiently by 'user' would be a non-starter.

**Review by snopoke — COMMENTED** [review]
General comment here. It seems we're putting a lot of effort (and some additional complexity) into making this a very general app when we only have a single use case currently. I get that it's good to publish tools that are generally useful but at least initially it seems unnecessary.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:13]: Addressed in a358992.

File: field_audit/field_audit.py
Comment (joxl) [reply field_audit/field_audit.py]: Changed to use a `ContextVar` in 5856369.

File: field_audit/models.py
Comment (snopoke) [inline field_audit/models.py:156]: what does this return for foreign key fields?

Comment (joxl) [reply field_audit/models.py:156]: It returns the "referenced" value. That is, the value of `<column>` in the example SQL snip: `FOREIGN KEY REFERENCES <table>(<column>)`.

This is often the primary key of the "referenced" record, but foreign keys don't have to reference a primary key, and this is how you get that value regardless of what the `REFERENCES` column is.

File: field_audit/models.py
Comment (joxl) [reply field_audit/models.py:130]: Added in 25400ab.

**Comment by dannyroberts** [comment]
Simon, Dennis, Kai, and I synced this morning, and I want to share the outcome and open this back up for review. With SaaS's goal being to have a way to start collecting generic model level audit data quickly and a Q3 resourcing constraint that will pause further big investment in our auditing framework, we agreed to pursue a strategy of getting what we have in now to begin collecting data, with the understanding that SaaS will continue to own the problem of any difficulties querying the data down the road.

So let's open this back up for review. As always, all manner of review is welcome. To help facilitate getting in now what's safe to get in, I'd ask reviewers to primarily consider immediate safety and correctness when giving blocking feedback, and to continue to give feedback about potential issues down the line but to note them as non-blocking feedback.

**Review by millerdev — APPROVED** [review]

File: field_audit/field_audit.py
Comment (snopoke) [inline field_audit/field_audit.py:38]: not blocking (and I can't recall if this was already discussed) but can you say why you chose to wrap the methods instead of using Django signals?

Comment (joxl) [reply field_audit/field_audit.py:38]: Sure. One of the reasons is because `_decorate_db_write` needs to perform logic before and after the write method is called, and using signals would have required splitting that logic into separate signal listeners (`pre_<event>` and `post_<event>`). In other words this really is use case for decorators because Django's signals framework would have fragmented the logic and made this code significantly more complex.

**Comment by joxl** [comment]
Waiting for feedback form @gherceg and @AmitPhulera yet.

**Review by AmitPhulera — APPROVED** [review]

**Review by gherceg — APPROVED** [review]

File: field_audit/field_audit.py
Comment (joxl) [reply field_audit/field_audit.py:89]: See #3

**Pull request merged**
